### PR TITLE
GraphQL query customisation

### PR DESCRIPTION
**Pull request by unclecheese**
Resolves https://github.com/silverstripe/silverstripe-graphql/issues/13

### CODE REVIEW

File: client/src/lib/dependency-injection/InjectorProvider.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import React from 'react';
+import provideInjector from './provideInjector';
+
+export default provideInjector((props, context) => <div>{children}</div>);
```
Comment (flamerohr) [inline client/src/lib/dependency-injection/InjectorProvider.js]: where's `children` defined? ;)

Comment (unclecheese) [reply client/src/lib/dependency-injection/InjectorProvider.js]: Not sure how this got in here. Should have been deleted.

File: client/src/lib/dependency-injection/buildApolloGraphqlContainer.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  /**
+   * A register of templates registered with this container
+   */
+  templates: {
+    [CREATE]: buildCreateMutation(captureTag),
+    [READ]: buildReadQuery(captureTag),
+    [READ_ONE]: buildReadOneQuery(captureTag),
+    [UPDATE]: buildUpdateMutation(captureTag),
+    [DELETE]: buildDeleteMutation(captureTag),
+    [QUERY]: buildBaseQuery(captureTag),
```
Comment (unclecheese) [inline client/src/lib/dependency-injection/buildApolloGraphqlContainer.js]: We should initialise with these empty and inflate them at boot time. Would help to decouple Injector from SilverStripe. Each consumer of Injector should in theory be supplying its own set of templates.

File: client/src/boot/buildApolloClient.js [hunk 3]
```diff
@@ -16,8 +16,8 @@ function buildApolloClient(baseUrl) {
     shouldBatch: true,
     addTypename: true,
     dataIdFromObject: (o) => {
-      if (o.id >= 0 && o.__typename) {
-        return `${o.__typename}:${o.id}`;
+      if ((o.id >= 0 || o.ID >= 0) && o.__typename) {
```
Comment (tractorcow) [inline client/src/boot/buildApolloClient.js]: Can you explain the case issue here? Should we check that at least one of id or ID are not undefined first?

Maybe shift the ID into a var instead for clarity.

File: client/src/boot/buildApolloClient.js [hunk 4]
```diff
@@ -16,8 +16,8 @@ function buildApolloClient(baseUrl) {
...
+        return `${o.__typename}:${(o.id >= 0) ? o.id : o.ID}`;
```
Comment (tractorcow) [inline client/src/boot/buildApolloClient.js]: Maybe we should shift id into a variable here for readability.

File: client/src/lib/dependency-injection/ApolloGraphqlManager.js [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    this.fragments = { ...fragments } || {};
+
+    this.reduceApolloConfig = this.reduceApolloConfig.bind(this);
+  }
+
+  setConfig(name, value) {
+    // prevent overriding functionally important configs
+    if (protectedConfig.includes(name)) {
+      throw new Error(`
+Tried to set protected config values: '${name}', which is discouraged.
```
Comment (tractorcow) [inline client/src/lib/dependency-injection/ApolloGraphqlManager.js]: Weird newline here.

File: client/src/lib/dependency-injection/ApolloGraphqlManager.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    }
+    this.config = {
+      ...this.config,
+      [name]: value,
+    };
+
+    return this;
+  }
+
+  transformApolloConfig(config, callback) {
```
Comment (tractorcow) [inline client/src/lib/dependency-injection/ApolloGraphqlManager.js]: jsdocs please :D

File: client/src/lib/dependency-injection/ApolloGraphqlManager.js [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+      [config]: [
+        ...transformList,
+        callback,
+      ],
+    };
+
+    return this;
+  }
+
+  addParam(param) {
```
Comment (tractorcow) [inline client/src/lib/dependency-injection/ApolloGraphqlManager.js]: jsdocs please :D

File: client/src/lib/dependency-injection/buildApolloGraphqlContainer.js (hunk 2 above)
Comment (tractorcow) [reply client/src/lib/dependency-injection/buildApolloGraphqlContainer.js]: Can you implement this or should we leave it for this merge @unclecheese ?

**Comment by tractorcow** [comment]
I think taking the asset-admin changes https://github.com/silverstripe/silverstripe-asset-admin/pull/753/files and constructing an equivalent unit test from this is sufficient. :) but with an arbitrary dataobject type.

File: client/src/lib/dependency-injection/buildApolloGraphqlContainer.js (hunk 2 above)
Comment (unclecheese) [reply client/src/lib/dependency-injection/buildApolloGraphqlContainer.js]: done

File: client/src/lib/dependency-injection/ApolloGraphqlManager.js (hunk 5 above)
Comment (flamerohr) [reply client/src/lib/dependency-injection/ApolloGraphqlManager.js]: I think all error messages should use something like this https://github.com/airbnb/deline to get rid of coding indents when it appears in the console

Those lengthy error messages plus indents is quite an eyesore now :D 

I can raise a pull request later on.

**Pull request merged**